File: prog3/List.java
//Ryan Brounley
//CMPS 101

import java.util.*;

public class List{

  private class Node{
	Object data;
	Node prev;
	Node next;
	//Constructor
	Node(Object data){
		this.data = data;
		next = null; 
		prev = null;
	}
	public String toString() { 
		return String.valueOf(data); }
  }

  private Node front;
  private Node curr;
  private Node back;
  private int length;

  //Constructor
  List(){
	front = curr = back = null;
	length = 0;
  } 
	
  //Access functions
  //Returns length of the this
  int length(){
      if(front == null) return 0;
      else 
	  return this.length; 
  }

  //Returns the index of the cursor element in this list
  //or returns -1 if the cursor element is undefined.	
  int getIndex(){
    Node A = front;
    int count = 0;
    if( curr != null){
      if(curr.data == A.data){ return (count + 1); }
      while( A != null){
        count++;
        if(curr.data == A.next.data){
          return count; 
        }
        A = A.next;  
        }  
     }return -1;
  }
     
  //returns whether or not the cursor is on the List
  boolean offList(){
        if(this.length()== 0)return true;
        else if(this.curr == null)return true;
        else
            return false;
    }

  boolean isFront(){
        if(this.length() == 0) return false;
        else if(this.curr == front) return true;
        else   return false;
  }

  //Returns front element in this list.
  Object front(){
      Node A = front;
      if(this.length() > 0) return A.data;
      else 
	throw new 
        RuntimeException("this Error: front() called on empty list");
  }
 
  //Return back element in this list.
  Object back(){ 
      Node B = back;
      if(this.length() > 0) return B.data;
      else
	  throw new 
          RuntimeException("this Error: back() called on empty list");
  }

  //returns cursor element in this list
  //Pre: length()>0, getIndex()>0
  Object getElement(){
      if((this.getIndex() > -1) && (this.length() > 0)) return curr.data;
      else
	  throw new 
          RuntimeException("this Error: getElement() called on empty list");
  }

  //Returns true if this List and L are the same integer
  //sequence. 
  public boolean equals(List L){
      boolean equalsflag = true;
      Node A = this.front;
      Node B = L.front;

      if(this.length() == L.length()){
	  while(equalsflag && A!= null){
	      equalsflag = (A.data == B.data);
	      A = A.next;
	      B = B.next;
	  }
	  return equalsflag;}
      else{
	  return false;
      }
  }

  //Manipulation procedure
  //Re-sets the List to empty
  void clear(){
      if(front == null) return;
      else{
	  front.next = null;
	  front = null;
	  length = 0;
      }
  }

  //Moves the cursor to the element at index i, otherwise
  //the cursor becomes undefined.
  void moveTo(int i){
      Node cursor = front;
      if ((0 <= i) && (i <= (this.length() - 1))){
	 int newcounter = 0;
	 while(newcounter < i){
	 	cursor = cursor.next;
		newcounter++;
	 }
	 curr = cursor;
      } else {
	 curr = null;
      }
  } 

  //Moves the cursor one step toward the front of the list.	  
  void movePrev(){ 
     Node A;
     if((this.length() == 0)||(this.curr == null))
	throw new 
        RuntimeException("this Error: movePrev() called on empty list");
     else{ 
	A = curr;
	curr = A.prev;
      }
  }

  //Moves the cursor one step toward the back of the list.
  void moveNext(){
     //Node A;
     if(this.length() == 0)
	throw new 
        RuntimeException("this Error: moveNext() called on empty list");
     else{
	if(curr != null)
		curr = curr.next;
	else
		System.out.println("curr was null!!");
	//A = this.curr;
	//this.curr = A.next;
	}
  }

  //Inserts new element before front element in the List.
  void prepend(Object data){
      Node tmp;
      Node node = new Node(data);
      if(this.length() == 0)
	  front = back = node;
      else{
	  tmp = front;
	  tmp.prev = node;
          front = node;
	  front.next = tmp;
      }
      length++;
  }

  //Inserts new element after back element in the List.
  void append(Object data){
      Node tmp;
      Node node = new Node(data);
      if(this.length() == 0){ 
	  front = back = node;}
      else{
	  tmp = back;
	  tmp.next = node;
	  back = node;
	  back.prev = tmp;
      }
      this.length++;
  }

  //Inserts new element before cursor element into this List.
  void insertBefore(Object data){
	if(this.length() == 0){
		throw new 
                RuntimeException("this Error: insertBefore() called on empty list");
	}
	length++;
	Node before = new Node(data);
	before.next = curr;
	if(curr != front){
		curr.prev.next = before;
		before.prev = curr.next;
	}
	curr.prev = before;
	if(curr == back) front = before;
  }

  //Inserts new element before cursor element in this List.
  void insertAfter(Object data){
	if(this.length() == 0){
		throw new 
                RuntimeException("this Error: insertAfter() called on empty list");
	}
	length++;
	Node after = new Node(data);
	after.prev = curr;
	if(curr != back){
		curr.next.prev = after;
		after.next = curr.next;
	}
	curr.next = after;

	if(curr == back) back = after;
  }
 
  //Deletes the front element of this List.
  void deleteFront(){
      if(this.length() == 0) 
	  throw new 
	  RuntimeException("this Error: deleteFront() called on empty this"); 
      if(this.length() > 1){
	  front = front.next;
	  front.prev = null;}
      else{
	  front = back = null;}
      length--;
  }
	
  //Deletes the back element of this List.
  void deleteBack(){
      if(this.length() == 0) 
	  throw new NoSuchElementException();
      else{
	  back = back.prev;
          back.next = null;
          }
  }

  //Destes cursor element from this List. Cursor is undefined after this.
  void delete(){
      if(this.length() <= 0 && this.getIndex() < 0) 
	  throw new NoSuchElementException();
      else{
	  Node temp = curr.prev;
	  temp.next = curr.next;
	  if(temp != null){
	      curr = curr.next;
	      curr.prev = temp;
	  }
      }
  }

   void deletePrev(){
      if(this.length() <= 0 && this.getIndex() < 0) 
	  throw new NoSuchElementException();
      else{
          if(curr.prev != null){
	     Node temp = curr.prev.prev;
	     temp.next = curr;
	     if(temp != null){
	        curr.prev = temp;
		curr.prev.prev.next = curr;
	     }else this.deleteFront();
      }
      else{ this.deleteFront();}
  }
  }
 

  //Other methods
  //Overrides toString method in Node class
      public String toString(){
	  String str = "";
	  for(Node N = front; N != null; N = N.next){
	      str += N.toString() + " ";
	  }
	  return str;
      }
}

File: prog5/List.c
//Ryan Brounley
//CMPS 101 Professor Tantalo

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <assert.h>
#include <stdbool.h>
#include "List.h"

typedef struct NodeObj{
     int data;
     struct NodeObj *prev; 
     struct NodeObj *next;
} NodeObj;

typedef struct NodeObj* Node;

typedef struct ListObj{
     Node front;
     Node curr;
     Node back;
     int length;
} ListObj;

//creates a new node with NodeObj qualities
Node newNode(int data){
   Node N = malloc(sizeof(NodeObj));
   N->data = data;
   N->next = NULL;
   return(N);
}

// // Frees heap memory pointed to by *pN, sets *pN to NULL.
// // Private.
void freeNode(Node* pN){
    if( pN!=NULL && *pN!=NULL ){
          free(*pN);
          *pN = NULL;
     }
}

//creates a new ListObj
List newList(void){
     List L = malloc(sizeof(ListObj));
     L->front = L->curr = L->back = NULL;
     L->length = 0;
     return L;
}
void freeList(List *pL) {
	if(pL==NULL || *pL==NULL) {
		return;
	}
	clear(*pL);
	free(*pL);
	*pL = NULL;
}
// Access functions -----------------------------------------------------------

// Returns number of elements in this list.
int length(List L){
     if(L->front == NULL) return 0;
     return L->length;
} 

// Returns the index of the curr element in this list, or
// returns -1 if the curr element is undefined.
int getIndex(List L){
     Node a = L->front;
     int count = 0;
     if(L->curr != NULL){
       if(L->curr->data == a->data){ return (count + 1); }
       while(a != NULL){
          count++;
          if(L->curr->data == a->next->data){
             return count;
          }
          a = a->next;
          }
     }
     return -1;
}

//Return the front element
int front(List L){
    if(length(L) > 0) return L->front->data;
    else{
       return -1;}
}

//Returns the back element
int back(List L){
    if(length(L) > 0){ return L->back->data;
    }else{
       return -1;
    }
}

//Returns curr element on the list
int getElement(List L){
    if((getIndex(L) > -1) && (length(L) > 0)) return L->curr->data;
    else{  
       fprintf(stderr,"curr is off the List");
       return -1;
    }
}

//Checks to see if two lists are identical or different
int equals(List A, List B){
      bool equalsflag = true;
      Node a = A->front;
      Node b = B->front;
      while(equalsflag && a != NULL){
            equalsflag = (a->data == b->data);
            a = a->next;
            b = b->next;
          }
      return equalsflag;
}

// Manipulation procedures ----------------------------------------------------

// Re-sets this List to the empty state.
void clear(List L){
    if(L->front == NULL){
//		fprintf(stderr, "Clearing empty list clear()");
		exit(1);}	
    while(L->front != NULL){
          deleteFront(L);
        }
    L->front = L->back = L->curr = NULL;
    L->length = 0;
}


// If 0<=i<=length()-1, moves the curr to the element
// at index i, otherwise the curr becomes undefined.
void moveTo(List L, int i){
    Node curr = L->front;
    if((0 <= i) && (i <= (length(L) - 1))){
       int newcounter = 0;
       while(newcounter < i){
            curr = curr->next;
            newcounter++;
       }
       L->curr = curr;
       } else {
         L->curr = NULL;
       }
} 

// Moves the curr one step towards the front of the list
void movePrev(List L){
    if((length(L) == 0)||(L->curr == NULL)){
        //opposite if for the preconditions^
        fprintf(stderr,"Empty list movePrev()");
        exit(EXIT_FAILURE);
    }else{
      Node a;
      a = L->curr;
      L->curr = a->prev;
    }
}

// Moves the curr one step toward the back of the list
void moveNext(List L){
     if(length(L) == 0){ 
        fprintf(stderr,"empty List moveNext()");
        exit(EXIT_FAILURE);
     }else{
        Node a;
        a = L->curr;
        L->curr = a->next;
     }
}

//inserts element to front  of the list
void prepend(List L, int data){ 
     if(L == NULL){
        fprintf(stderr,"error L == NULL prepend");
        return;}
     Node n = malloc(sizeof(struct NodeObj));
     n->data = data;
     n->prev = NULL;
     n->next = L->front;
     if(L->front == NULL){
        L->back = n;
     }else{
        L->front->prev = n;
     }
     L->front = n;
     L->length++;
}

//inserts element to back fo the list
void append(List L, int data){
    if(L == NULL){
        fprintf(stderr,"error L == NULL append\n");
        return;}
     Node n = malloc(sizeof(struct NodeObj));
     n->data = data;
     n->next = NULL;
     n->prev = L->back;
     if(L->back == NULL){
        L->front = n;
    }else{
        L->back->next = n;
    }
    L->back = n;
    L->length++;
}
 
//Inserts new element before the curr element
void insertBefore(List L, int data){
  
   if(length(L) != 0 && getIndex(L) != -1 ){
      if (getIndex(L) == 1){
         prepend(L, data);
         return;
      }
      else{
         Node before = newNode(data);
         L->curr->prev->next = before;
         before->prev = L->curr->prev;
         before->next = L->curr;
         L->curr->prev = before;
      }
      L->length++;
   }
} 

//inserts new element after the curr element
void insertAfter(List L, int data){
     if(length(L) == 0){
          fprintf(stderr, "empty List InsertAfter()");
          exit(EXIT_FAILURE);}
     L->length++;
     Node after = newNode(data);
     after->prev = L->curr;
     if(L->curr != L->back){
           L->curr->next->prev = after;
           after->next = L->curr->next;
     }
     L->curr->next = after;
     if(L->curr == L->back) L->back = after;
}

//Deletes the front element in this list
void deleteFront(List L){
     Node tmp = NULL;
     if(length(L) == 0){
            fprintf(stderr,"empty list deleteFront()");
            exit(EXIT_FAILURE);}
     else if(L == NULL){
            fprintf(stderr,"empty list deleteFront()");
            exit(EXIT_FAILURE);}
     else{
        tmp = L->front;
	if(length(L) > 1) {
		L->front = L->front->next;
		L->front->prev = NULL;
        }else {L->front = L->back = NULL;}
     }
     L->length--;
     freeNode(&tmp);
}

//Deletes the back element in this list
void deleteBack(List L){ 
     Node tmp = NULL;
     if(length(L) == 0){
            fprintf(stderr,"empty list deleteFront()");
            exit(EXIT_FAILURE);}
     else if(L == NULL){
            fprintf(stderr,"empty list deleteFront()");
            exit(EXIT_FAILURE);}
     else{
        tmp = L->back;
	if(length(L) > 1) {
		L->back = L->back->prev;
		L->back->next = NULL;
        }else {L->front = L->back = NULL;}
     }
     L->length--;
     freeNode(&tmp);
}
//Deletes curr element in this list
void delete(List L){
     Node tmp;
     if(length(L) <= 0 && getIndex(L) < 0)
        exit(EXIT_FAILURE);
     else{
        tmp = L->curr->prev;
        tmp->next = L->curr->next;
        if(tmp != NULL){
           L->curr = L->curr->next;
           L->curr->prev = tmp;
        }
     }
}

// Other operations -----------------------------------------------------------

//prints the list out
void printList(FILE* out, List L){
    if(L == NULL){
       fprintf(stderr,"Empty List printlist()");
       exit(EXIT_FAILURE);
    }
    if(out == NULL){
       fprintf(stderr,"no out FILE printlist()");
       return;
    }
    Node n = L->front;
    while(n != NULL){
         fprintf(out, "%d ",n->data);
         n = n->next;
    }
}

//copys the input list and returns it into a new list.
List copyList(List L){
    List cpy = newList();
    Node A = L->front;
    while(A != NULL){
        append(cpy, A->data);
        A = A->next;
    }
    return cpy;
}

